[PyTorch][Distributed] Integrate with ShardedOptimizer in the unit test of ShardedLinear

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69569
* #69725

Since ShardedOptimizer is added in https://github.com/pytorch/pytorch/pull/68607. We now integrate it in our unit test for Sharded Linear.

Differential Revision: [D32777020](https://our.internmc.facebook.com/intern/diff/D32777020/)

Commit:
[PyTorch][Distributed] Integrate with ShardedOptimizer in the unit test of ShardedLinear

Since ShardedOptimizer is added in https://github.com/pytorch/pytorch/pull/68607. We now integrate it in our unit test for Sharded Linear.

Differential Revision: [D32777020](https://our.internmc.facebook.com/intern/diff/D32777020/)

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/_sharded_tensor/ops/test_linear.py:
```diff
@@ -84,14 +88,28 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
+            sharded_dim, start_pos, chunk_size
+        )
+        self.assertEqual(sharded_weight.size(), local_weight_narrowed.size())
+        with torch.no_grad():
```
Why do we need `torch.no_grad` here?

fduwjj replied:
It complains if not added. Because here `previous_sharded` is from detach() which does not come with gradient. And looks like the result from narrow also does not have gradient.

pritamdamania87 commented on test/distributed/_sharded_tensor/ops/test_linear.py:
```diff
@@ -84,14 +88,28 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
+        self.assertEqual(sharded_weight.size(), local_weight_narrowed.size())
+        with torch.no_grad():
+            self.assertNotEqual(previous_sharded, sharded_weight)
+            self.assertEqual(sharded_weight, local_weight_narrowed)
```
Shouldn't we verify the bias as well?

fduwjj replied:
sure.

Commit:
Update on "[PyTorch][Distributed] Integrate with ShardedOptimizer in the unit test of ShardedLinear"


Since ShardedOptimizer is added in https://github.com/pytorch/pytorch/pull/68607. We now integrate it in our unit test for Sharded Linear.

Differential Revision: [D32777020](https://our.internmc.facebook.com/intern/diff/D32777020/)

[ghstack-poisoned]

fduwjj commented:
Addressed the feedback from reviewers. This PR is dependent on https://github.com/pytorch/pytorch/pull/69725 and https://github.com/pytorch/pytorch/pull/68607. The test might fail for now but I patch dependent and test locally, the test passes.

wanchaol reviewed (approved):
lgtm, thanks!

Commit:
Update on "[PyTorch][Distributed] Integrate with ShardedOptimizer in the unit test of ShardedLinear"


Since ShardedOptimizer is added in https://github.com/pytorch/pytorch/pull/68607. We now integrate it in our unit test for Sharded Linear.

Differential Revision: [D32777020](https://our.internmc.facebook.com/intern/diff/D32777020/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Distributed] Integrate with ShardedOptimizer in the unit test of ShardedLinear"


Since ShardedOptimizer is added in https://github.com/pytorch/pytorch/pull/68607. We now integrate it in our unit test for Sharded Linear.

Differential Revision: [D32777020](https://our.internmc.facebook.com/intern/diff/D32777020/)

[ghstack-poisoned]
